vm/ubpf_run.c

```diff
@@ -21,29 +21,133 @@
 //#include <stdbool.h>
 //#include <stdarg.h>
 //#include <inttypes.h>
-//#include <sys/mman.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
 #include "ubpf_int.h"
 
+static void enable_instruction_count(int fd);
+static void disable_instruction_count(int fd);
+static int setup_instruction_counter(void);
+static long long get_instruction_count(int fd);
+
 uint64_t ubpf_run(struct ubpf_vm *vm, void *mem, size_t mem_len, bool jit,
 	struct ubpf_inst_cnt *inst_cnt)
 {
-	uint64_t ret;
-	char *errmsg;
-
-	if (jit) {
-		ubpf_jit_fn fn = ubpf_compile(vm, &errmsg);
-		if (fn == NULL) {
-			fprintf(stderr, "Failed to compile: %s\n", errmsg);
-			free(errmsg);
-			return 1;
-		}
-		ret = fn(mem, mem_len);
-	} else {
-		if (ubpf_exec(vm, mem, mem_len, &ret, inst_cnt) < 0)
-			ret = UINT64_MAX;
+    uint64_t ret;
+    char *errmsg;
+    int cntrfd;
+    ubpf_jit_fn fn;
+
+    cntrfd = setup_instruction_counter();
```

Comment:
You should check if instruction count has been enabled before calling this. As it is, running on non-Linux will always fail on setup_instruction_counter().